<img src="https://drive.google.com/uc?id=1E_GYlzeV8zomWYNBpQk0i00XcZjhoy3S" width="100"/>  
# Drill Week 3: Feature Manipulation   

In this drill, you will use essential pandas manipulations on a NEW dataset.   
We will get this dataset from an outside source :)    

Specifically, we will be using the wine dataset.  


<img src="https://img.clipartlook.com/wine-clipart-set-of-vector-and-watercolor-drawings-of-wine-glasses-vector-art-illustration-612.jpg" width="400"/>  


In [ ]:
from sklearn.datasets import load_wine

import pandas as pd  # fill in with the proper package name 


This dataset contains 3 types of wines, and the following measures for each   

- Alcohol
- Malic acid
- Ash
- Alcalinity of ash
- Magnesium
- Total phenols
- Flavanoids
- Nonflavanoid phenols
- Proanthocyanins
- Color intensity
- Hue
- OD280/OD315 of diluted wines
- Proline

In [ ]:
data = load_wine() 
features = data.data # these are our features 
labels = data.target
df = pd.DataFrame(features)
df.columns = data.feature_names
df['labels'] = labels

In [ ]:
df

alcohol  malic_acid   ash  ...  od280/od315_of_diluted_wines  proline  labels
0      14.23        1.71  2.43  ...                          3.92   1065.0       0
1      13.20        1.78  2.14  ...                          3.40   1050.0       0
2      13.16        2.36  2.67  ...                          3.17   1185.0       0
3      14.37        1.95  2.50  ...                          3.45   1480.0       0
4      13.24        2.59  2.87  ...                          2.93    735.0       0
..       ...         ...   ...  ...                           ...      ...     ...
173    13.71        5.65  2.45  ...                          1.74    740.0       2
174    13.40        3.91  2.48  ...                          1.56    750.0       2
175    13.27        4.28  2.26  ...                          1.56    835.0       2
176    13.17        2.59  2.37  ...                          1.62    840.0       2
177    14.13        4.10  2.74  ...                          1.60    560.0       2

[178 rows x 14 columns]

## Drilling Feature Manipulation   
Use pd.columns to return the title of each feature 

Your equipment measurer was not callibrated correctly. Add 2 points to the magnesium of each entry 

Create a new feature called flav_to_non_flav ratio that is the ratio between flavanoids and nonflavanoid_phenols.   
Run a df.head to make sure your changes were made 

Create a new dataframe called high_alcohol_content. Have it only contain the entries with the alcohol value over 13.5  
Hint -- use the solution as shown in the lecture 

Increase the malic_acid value by 1 where the total_phenols is over 1.7

## Drilling Feature Selection   
You hate the number 13. Use loc to drop the entry at position 13. (Remember -- Python uses 0 indexing!) 

You also know that having too many features may increase computational overhead and not help accuracy.  
Drop the 3 columns you think are the least helpful in determining the wine class 

#Preprocessing

For each column, find the number of missing values

Take the correlation among the features using a heatmap. Drop the 3 columns with the lowest amount of correlation with "label" in magnitude.

## Drilling Critical Thinking    
Say you are given a new dataset. You see there are over 30 features.   
What are some benefits and drawbacks to having this many features?     
If you had to drop some, what are a few ways that you could select which features to keep?  